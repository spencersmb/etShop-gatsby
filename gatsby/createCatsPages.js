const path = require("path")

module.exports = async ({ actions, graphql }) => {
  const { createPage } = actions
  const perPage = 10
  const allCats = {}
  const excludeItems = ["uncategorized"]
  let pageNumber = 1

  const fetchCatPages = async (variables) =>
    await graphql(GET_CATS, variables).then(async ({ data }) => {
      const {
        wpgraphql: {
          categories: {
            nodes
          }
        }
      } = data

      await Promise.all(nodes
        .filter(node => !excludeItems.includes(node.slug))
        .map((cat) => {
          const {
            supportQuestions: {
              pageInfo: { hasNextPage, endCursor, hasPreviousPage }
            }
          } = cat
          const questions = cat.supportQuestions.nodes
          const catPageTemplate = path.resolve(`./src/templates/supportCatListPage.tsx`)
          const catPagePath = `/support/category/${cat.slug}/page/${pageNumber}`
          // console.log("cat.supportQuestions.pageInfo", cat.supportQuestions.pageInfo)
          // console.log("slug", cat.slug)
          // console.log("ID", cat.id)
          // console.log("pageNumber", pageNumber)
          // console.log("hasNextPage", hasNextPage)
          // console.log("hasPrevPage", hasPreviousPage)

          // add item to be excluded if no next page
          if (!hasNextPage) {
            excludeItems.push(cat.slug)
          }
          const prevPages = allCats[cat.slug] && allCats[cat.slug].pages ? allCats[cat.slug].pages : {}
          // check if object exists first
          allCats[cat.slug] = {
            pages: {
              ...prevPages,
              [pageNumber]: {
                path: catPagePath,
                component: catPageTemplate,
                context: {
                  cat: cat.name,
                  catId: cat.id,
                  questions,
                  pageNumber,
                  hasNextPage,
                  hasPrevPage: hasPreviousPage
                }
              }
            }
          }

          if (hasNextPage) {
            pageNumber++
            return fetchCatPages({ first: perPage, after: endCursor })
          }
        }))
      return allCats
    })

  await fetchCatPages({ first: perPage, after: null }).then(catItems => {
    Object.keys(catItems).map((cat) => {
      const pages = Object.keys(catItems[cat].pages)
      pages.map(page => {
        console.log(`create Cat Page ${cat}`)
        console.log(`create Cat PageNumber ${catItems[cat].pages[page].context.pageNumber}`)
        createPage(catItems[cat].pages[page])
      })
    })
  })
}

const GET_CATS = `
  query GET_ALL_CATS($first:Int $after:String){
      wpgraphql{
        categories {
          nodes {
            id
            name
            slug
            supportQuestions(after: $after, first: $first) {
              pageInfo{
                  hasNextPage
                  hasPreviousPage
                  endCursor
              }
              nodes {
                  title
                  id
                  slug
                  excerpt
                  acfSupportQuestions{
                      popularity
                  }
              }
            }
          }
        }
      }
    }
  `
