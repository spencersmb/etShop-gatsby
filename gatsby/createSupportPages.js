const path = require(`path`)
module.exports = async ({ actions, graphql }) => {
  const supportQuery = `
  {
  wpgraphql {
    supportQuestions(where: {status: PUBLISH}, first: 100){
      nodes{
        date
        uri
        id
        slug
        title
        content
        featuredImage {
          altText
          caption
          sourceUrl
        }
      }
    }
  }
}
  `

  return new Promise((resolve, reject) => {
    graphql(supportQuery).then(results => {

      if (results.errors) {
        console.log(results.errors)
        reject(results.errors)
      }
      results.data.wpgraphql.supportQuestions.nodes.forEach(question => {
        actions.createPage({
          path: `/support/${question.slug}/`,
          component: path.resolve(`./src/templates/supportQuestion.tsx`),
          context: {
            // blogId: blog.id,
            ...question
          }
        })
        console.log(`Created page: /support/${question.slug}`)
      })
      resolve()
    })

  })
}
