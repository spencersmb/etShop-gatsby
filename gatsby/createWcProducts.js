const path = require("path")
const GET_PAGES = `
  query GET_WC_PAGES{
      allWcProduct{
        edges{
          node{
            slug
            name
          }
        }
      }
    }
  `
module.exports = async ({ actions, graphql }) => {
  const { createPage } = actions
  const allPages = []
  const fetchPages = async () =>
    await graphql(GET_PAGES).then(({ data }) => {
      const {
        allWcProduct: {
          edges
        }
      } = data
      edges.map(page => {
        allPages.push(page.node)
      })
      return allPages
    })

  await fetchPages().then(allPages => {
    allPages.map(page => {
      console.log(`create page: ${page.slug}`)
      createPage({
        path: `/products/${page.slug}`,
        component: path.resolve(`./src/templates/productPage.tsx`),
        context: {
          slug: page.slug
        }
      })
    })
  })

  // return new Promise((resolve, reject) => {
  //   graphql(pageQuery).then(results => {
  //
  //     if (results.errors) {
  //       console.log(results.errors)
  //       reject(results.errors)
  //     }
  //     results.data.allWcProduct.edges.forEach(({ node }) => {
  //       createPage({
  //         path: `/products/${node.slug}`,
  //         component: path.resolve(`./src/components/products/productDetailPage.tsx`),
  //         context: {
  //           slug: node.slug
  //         }
  //       })
  //       console.log(`Created page: /products/${node.slug}`)
  //     })
  //   })
  //   resolve()
  // })
}
