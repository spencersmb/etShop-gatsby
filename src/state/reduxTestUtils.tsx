// this is a handy function that I normally make available for all my tests
import { ICartState } from '@et/types/Cart'
import { IGatsbyConfig } from '@et/types/Gatsby'
import { IProduct } from '@et/types/Products'
import React from 'react'
import initialState from '@redux/reducers/initialState'
import { Provider } from 'react-redux'
import { render } from 'react-testing-library'
// that deal with connected components.
// you can provide initialState or the entire store that the ui is rendered with
import { createStore } from 'redux'

export function renderWithRedux (
	ui: any,
	reducer: any,
	{ state, store = createStore(reducer, state) }: any = initialState
) {
	return {
		...render(<Provider store={store}>{ui}</Provider>),
		// adding `store` to the returned utilities to allow us
		// to reference it in our tests (just try to avoid using
		// this to test implementation details).
		store
	}
}

const testSiteMetaDataQuery = {
	title: `Every-Tuesday Digital Products Shop`,
	description: `Kick off your next, great Gatsby project with this default starter. This barebones starter ships with the main Gatsby configuration files you might need.`,
	author: `@Teelac`,
	authorUrl: 'https://every-tuesday.com/about/#teela',
	siteUrl: `https://shop.every-tuesday.com`,
	siteName: `Every-Tuesday Shop`,
	db: 'http://shopeverytuesday.local',
	route: 'et-shop',
	twitterUrl: 'https://twitter.com/teelacunningham',
	twitterDefaultImage: 'url'
}

export const singleItemQuery: IGatsbyConfig & { wcProduct: IProduct } = {
	site: {
		siteMetadata: {
			...testSiteMetaDataQuery
		}
	},
	wcProduct: {
		type: 'simple',
		date_created_gmt: '2018-09-25T20:35:51',
		date_modified_gmt: '2019-02-18T21:44:25',
		id: '202eca74-fc90-56e7-8269-b59f18a19194',
		name: 'Watercolor texture kit Vol. 1',
		price: '16',
		product_id: 222,
		pwyw: false,
		regular_price: '16',
		sale_price: '',
		slug: 'watercolor-texture-kit-vol-1',
		description: 'description',
		short_description: 'short desc',
		on_sale: false,
		features: {
			description: 'feature desc',
			items: {
				description: 'item desc',
				icon: 'icon',
				title: 'singleItemQuery title'
			}
		},
		categories: [
			{ id: 1, slug: 'fonts', name: 'Fonts' },
			{ id: 2, slug: 'Watercolor', name: 'Watercolor' }
		],
		images: [
			{
				id: 123414,
				alt: 'alt',
				fullSize: {
					url: 'fullsize-url'
				},
				thumbnail: {
					url: 'image-url'
				}
			}
		],
		license: {
			hasExtendedLicense: true,
			type: 'standard',
			extendedItem: {
				slug: 'watercolor-texture-kit-vol-1-ext'
			}
		},
		seo: {
			desc: 'Add serious vibrant color and detail with this kit packed with 32 unique watercolor textures, 12 seamless, repeatable watercolor texture patterns and 2 bonus watercolor paper patterns.',
			title: 'Watercolor texture kit Vol. 1'
		},
		tags: [
			{ id: 1, name: 'Texture Kit', slug: 'texture-kit' },
			{ id: 2, name: 'Waterbrush', slug: 'waterbrush' }
		],
		localFile: {
			name: 'file name',
			id: '1234',
			childImageSharp: {
				fluid: {
					src: 'src'
				},
				fixed: {
					width: '400',
					height: '400',
					src: 'src'
				}
			}
		}
	}
}

export const testCartEmpty: ICartState = {
	couponCode: {
		code: 'string',
		discount: 'string',
		loading: false,
		product_ids: [],
		submitted: false,
		type: 'string',
		valid: false,
	},
	isOpen: false,
	items: {
	},
	loaded: true,
	paymentType: 'Stripe',
	totalItems: 0,
	totalPrice: 0
}
