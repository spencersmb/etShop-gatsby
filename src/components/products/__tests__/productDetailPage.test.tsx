import React from 'react'
import renderer from 'react-test-renderer'
import ProductPage from '../productDetailPage'
import { StaticQuery } from 'gatsby'
import { singleItemQuery } from '@redux/reduxTestUtils'

beforeEach(() => {
	// @ts-ignore
	StaticQuery.mockImplementationOnce((mock) =>
		mock.render(singleItemQuery)
	)
})

describe('Product Layout', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(<ProductPage data={singleItemQuery}/>)
			.toJSON()
		expect(tree).toMatchSnapshot()
	})
})