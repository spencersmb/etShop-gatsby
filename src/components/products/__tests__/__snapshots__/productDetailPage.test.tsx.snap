// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Layout renders correctly 1`] = `
<h1>
  Item
</h1>
`;
